src/tomocupy_stream/rec.py

```diff
@@ -49,70 +49,220 @@
+        self.item_gpu['flat'] = cp.ones([2, *self.shape_flat_chunk], dtype=in_dtype)
 
         # pinned memory for reconstrution
-        self.rec_pinned = utils.pinned_array(np.zeros([16,*self.shape_recon_chunk], dtype=args.dtype))
+        self.rec_pinned = utils.pinned_array(np.zeros([16,*self.shape_recon_chunk], dtype=dtype))
 
         # gpu memory for reconstrution
-        self.rec_gpu = cp.zeros([2, *self.shape_recon_chunk], dtype=args.dtype)
+        self.rec_gpu = cp.zeros([2, *self.shape_recon_chunk], dtype=dtype)
                 
         # chunks        
-        self.nzchunk = int(np.ceil(args.nz/args.ncz))
+        self.nzchunk = int(np.ceil(nz/ncz))
         self.lzchunk = np.minimum(
-            args.ncz, np.int32(args.nz-np.arange(self.nzchunk)*args.ncz))  # chunk sizes        
-        self.ncz = args.ncz
-        self.nz = args.nz
+            ncz, np.int32(nz-np.arange(self.nzchunk)*ncz))  # chunk sizes        
+        self.ncz = ncz
+        self.nz = nz
         
         # streams for overlapping data transfers with computations
         self.stream1 = cp.cuda.Stream(non_blocking=False)
         self.stream2 = cp.cuda.Stream(non_blocking=False)
         self.stream3 = cp.cuda.Stream(non_blocking=False)
         
-        self.cl_tomo_func = tomo_functions.TomoFunctions(args,theta)
-        
         # threads for filling the resulting array
         self.write_threads = []
         for k in range(16):#16 is probably enough but can be changed
             self.write_threads.append(utils.WRThread())
         
-    def recon_all(self, result, data_in, dark_in, flat_in):
+    def recon_all(self, data, dark, flat, theta, output=None):
+        # Validate that the inputs match what was declared in __init__ and pre-allocated for.
+        expected_data_shape = (self.nz, self.nproj, self.n)
+        if data.shape != expected_data_shape:
+            raise ValueError(f"Expected data.shape {expected_data_shape}, got {data.shape}")
+        expected_dark_shape = (self.ndark, self.nz, self.n)
+        if dark.shape != expected_dark_shape:
+            raise ValueError(f"Expected dark.shape {expected_dark_shape}, got {dark.shape}")
+        expected_flat_shape = (self.nflat, self.nz, self.n)
+        if flat.shape != expected_flat_shape:
+            raise ValueError(f"Expected flat.shape {expected_flat_shape}, got {flat.shape}")
+        if output is None:
+            # Allocate output array.
+            output = np.zeros([self.nz, self.n, self.n], dtype=self.dtype)
+        else:
+            # Use pre-allocated output array, first validating it.
+            expected_output_shape = (self.nz, self.n, self.n)
+            if output.shape != expected_output_shape:
+                raise ValueError(f"Expected output.shape {expected_output_shape}, got {output.shape}")
                 
+        self.cl_tomo_func = tomo_functions.TomoFunctions(theta=theta, **self.tomofunc_kwargs)
```

Comment:
this class involves the initialization of gpu memory, precalculating interpolation grids and kernels, which takes quite some time. The class should be created only once in __init__.py. For the streaming processing, we plan to run recon_all(data,..) for data arrays having the same shape. We dont need to spend time for precalculating all interpolation kernels/grids several times. So, if there is no alternatives, I would move this call back in __init__

Reply:
addressed by [this](https://github.com/nikitinvv/tomocupy-stream/commit/195c64cae1fcd744687cd1c1af2b52daf0dd2878)